### ISSUE TITLE
Log messages are lost when an exception is thrown

### ISSUE THREAD

**Issue by otbe** [open-telemetry/opentelemetry-java-instrumentation#2285]
Hello

**Describe the bug**
Im not quite sure if this is a bug of the java instrumentation or something else but I hope you can guide me somehow towards a solution.

We recently switched from Xray SDK to AWS ADOT via javaagent and sidecar collector. In general this works pretty good and a lot better than Xray itself but we noticed some issues especially when an exception is thrown. 
We have a small Spring Boot app that uses standard logback logging via [logstash-logback-encoder](https://github.com/logstash/logstash-logback-encoder).  It happens in special cases that logs are completely dropped when an exception is thrown deep in the call hierarchy. It doesn't happen for all exceptions, it seems to be related to the depth of call/trace. What I mean by log messages are dropped is that we cant see any error handling log output from the enclosing classes. It just seems to be stuck at some point. 

What I tried so far:
- disabled the logstash encoder stuff, basically switched back to standard out logging via logback -> same behaviour
- disabled javaagent -> logs come back
- adding a try/catch block somewhere in the middle around the problematic code -> logs come back
- set the agent to debug log -> no errors or warning around this time, seems quite normal

**Steps to reproduce**
To be honest I don't know how to produce this if its really related to the call depth. I can try if you want me to do it :)

**What did you expect to see?**
Logs are still shipped/printed.

**What did you see instead?**
No logs.

**What version are you using?**
v0.15.0 (re-packaged version for AWS https://github.com/aws-observability/aws-otel-java-instrumentation)
Spring Boot 2.4.2

**Environment**
Compiler: corretto 1.8
OS: debian

I know this is somehow a bad bug report, but I don't know where to look at.

Thank you very much.

**Comment by anuraaga** [comment]
Hi @otbe - that does sound like a bug but nothing is apparent at a first glance. To clarify, do you expect a spring exception handler to catch and log the exception? Or do you have another mechanism for logging it?

If you're able to, try disabling logback or spring instrumentation individually, we may be able to pinpoint problematic instrumentation

https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/docs/suppressing-instrumentation.md

`logback` or `spring-webmvc` (guessing you're using MVC?)

**Comment by otbe** [comment]
Yes we're using spring mvc.

Just to be sure `-Dotel.instrumentation.logback.enabled=false` should work in 0.15.0?

**Comment by anuraaga** [comment]
With all the versions it's hard for me to keep up with the flags too - but yeah that should work on 0.15.0 https://github.com/open-telemetry/opentelemetry-java-instrumentation/tree/v0.15.0#disabled-instrumentations

**Comment by otbe** [comment]
Okay I can confirm `-Dotel.instrumentation.logback.enabled=false` works, no traceId in our logs afterwards. 

I tried disabling `logback` and afterwards `logback` and `spring-webmvc` but it doesnt seem to work. Exceptions are still not shown.

**Comment by otbe** [comment]
I was able to setup a reproduction repo:
https://github.com/otbe/factcast/tree/adot-issue
 
After you build the project by running `mvn install -DskipTests` you should able to run the test `org.factcast.itests.factus.FactusClientTest.testSubscription`. The test will fail 
and log something like

```
Unhandled onError:
io.grpc.StatusRuntimeException: CANCELLED: Failed to read message.
...
Caused by: java.lang.IllegalStateException: broken
```

If you now add the javaagent (pure opentelemtry or aws flavoured one) to the test execution the test will just hang. So its more problematic than I thought because its not about logs getting dropped its more about the execution seems to hang for some reason. If you comment out the `IllegalStateException` [here](https://github.com/otbe/factcast/blob/adot-issue/factcast-itests/factcast-itests-factus/src/test/java/org/factcast/itests/factus/proj/SubscribedUserNames.java#L43) the test fails (due to assertion errors).

**Comment by anuraaga** [comment]
Thanks a lot for this repro! It appears to be a bug in our gRPC instrumentation - `-Dotel.instrumentation.grpc.enabled=false` seems to prevent the deadlock. I'll continue to check out what the root cause is and make sure it fixes this repro case.

**Comment by anuraaga** [comment]
@otbe The latest snapshot for this upstream agent should have the fix

https://oss.sonatype.org/content/repositories/snapshots/io/opentelemetry/javaagent/opentelemetry-javaagent/0.17.0-SNAPSHOT/

It's compatible with `opentelemetry-api` 0.16.0. A new aws distro will probably be released for 0.17.0 though we're still sorting that out.

**Comment by otbe** [comment]
Thank you very much! I think we will wait for the upcoming releases (especially the aws flavoured agent)
